expo/expo
# [docs] Upgrade to Next 12 and SWC

# Why

Closes ENG-5611

This is an update of https://github.com/expo/expo/pull/17297

This also seems to make the docs build process faster. Previously I had a PR that took [8m3s](https://github.com/expo/expo/runs/7291200521?check_suite_focus=true) to build, and this one took [3m48s](https://github.com/expo/expo/runs/7322490012?check_suite_focus=true).

# How

1. Remove the ESBuild integration, since we can now use SWC
2. Switch over to Webpack 5; Next 12 doesn't support 4 anymore
3. Only add Babel back for the preval mechanism in the **docs/constants/**** files
4. Enable and configure SWC for the docs, including for emotion styles
5. Add a resolution workaround for MDX v1 to use only 1 babel version

# Test Plan

- See if it builds in CI
- Run it locally with `yarn dev`
- Build it with `yarn build`, then run it locally with `yarn run next start`
- Check if the pages are loading properly, e.g.
  - API version pages
  - API app config page
  - Random guides
  - Random RN API pages
